### PR TITLE
Identify public & private groups for logged in users

### LINKED ISSUE #80: Re-title "Private Nextstrain Groups" section header on splash page or restrict tiles to private groups
@jameshadfield [brought this up in our #nextstrain-general Slack channel](https://bedfordlab.slack.com/archives/C7SDVPBLZ/p1578945475043700) when he realized he was seeing `blab` under Private Nextstrain Groups. 

In that Slack thread, @trvrb describes how he wants `blab` to exist as a public Nextstrain group. 

Ideally, we would show users their "Nextstrain Group Membership" including public & private with something to make it clear that links to private groups are only showing up because a user (with access) is logged in.

### PR DESCRIPTION
**Pull request by jameshadfield**
See commit messages for more details.

image

@tsibley can you please review b70246f as it adds some static methods to the `Source` class(es) and modifies the `visibleGroups` function.

@trvrb are you happy with the UI here? It would be simple to show this `Nextstrain Groups` section on the splash page for everyone (i.e. non logged in users), and this wouldn't be too bad with the small number of groups we currently have, but I don't think it should be above the core tiles. Let me know if you want me to do this.

Available for testing at https://nextstrain-dev.herokuapp.com/

Closes #80

### CODE REVIEW

File: auspice/server/sources.js [hunk 1]
```diff
@@ -26,6 +26,12 @@ class Source {
   get baseUrl() {
     throw InvalidSourceImplementation("baseUrl() must be implemented by subclasses");
   }
+  static isGroup() { /* is the source a "nextstrain group"? */
+    return false;
+  }
+  static isPrivate() { /* does the source require authentication? */
+    return false;
+  }
```
Comment (tsibley) [inline auspice/server/sources.js]: The `isPrivate()` methods are duplicative. Calling `visibleToUser()` or `visibleToUser(null)` is valid and used elsewhere in the codebase to see if a Source is visible to an anonymous user.

File: authn.js [hunk 2]
```diff
@@ -210,25 +210,33 @@ function setup(app) {
...
-  const visibleGroups = (user) => Array.from(sources.values())
-    .filter(source => source.visibleToUser(user))
-    .map(source => source._name)
-    .filter(source => user.groups.includes(source));
+  const visibleGroups = (user) => Array.from(sources)
+      .filter(([, source]) => source.isGroup())
+      .filter(([, source]) =>
+        source.isPrivate() ?
+          source.visibleToUser(user) :
+          true /* allow all public sources regardless of user data */
```
Comment (tsibley) [inline authn.js]: The use of `isPrivate()` here subverts the intent that `visibleToUser(…)` is the sole authorizer of a (anonymous or logged in) user's ability to see a Source. As noted above, the `isPrivate()` method is unnecessary, and I think it should be removed.

File: static-site/src/layouts/userDataWrapper.jsx [hunk 3]
```diff
@@ -11,7 +11,7 @@ export default class UserDataWrapper extends React.Component {
   }
 
   loadUser() {
-    console.log("Fetching user data");
+    // console.log("Fetching user data");
```
Comment (tsibley) [inline static-site/src/layouts/userDataWrapper.jsx]: Looks like an unrelated change made during dev slipped into this commit.

File: auspice/server/sources.js (hunk 1 above)
Comment (jameshadfield) [reply auspice/server/sources.js]: How can I use `visibleToUser` (or another method) to know whether a group (source) is visible to a user because it is public or because it's private and the user is authenticated? I could check for it being an instance of `PrivateGroupSource` but this seems more complex than the solution I implemented.

Comment (jameshadfield) [reply auspice/server/sources.js]: 👍 resolved

File: authn.js (hunk 2 above)
Comment (jameshadfield) [reply authn.js]: 👍 resolved

File: static-site/src/layouts/userDataWrapper.jsx (hunk 3 above)
Comment (jameshadfield) [reply static-site/src/layouts/userDataWrapper.jsx]: resolved

**Comment by jameshadfield** [comment]
👍 thanks @tsibley -- have resolved all requested changes.

---

> I think the addition of isGroup() conflates the concept of a Nextstrain Group (for which we don't have first class concepts/objects/records for in the codebase) with the Source concept. 

Our current implementation is that nextstrain groups _are_ sources (although not all sources are groups) and I think it's necessary for a source to be able to interrogate itself to know if it's a "nextstrain group" or not. From your answer it's clear to me that you've thought about this and have a clear plan for how we these concepts are going to be implemented in the future. Looking forward to discussing this plan in the future (if there are docs I must have missed them and would appreciate you pointing me to them).

**Comment by tsibley** [comment]
> From your answer it's clear to me that you've thought about this and have a clear plan for how we these concepts are going to be implemented in the future. Looking forward to discussing this plan in the future (if there are docs I must have missed them and would appreciate you pointing me to them).

Your intuition is correct, I do have a plan in my head. I look forward to (hopefully sooner than later) having time to put it down in writing!
